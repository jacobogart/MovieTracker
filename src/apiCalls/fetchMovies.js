import { jacobKey, baseUrl } from '../utils/API-logins'

export const fetchMovies = () => {
  fetch(`${baseUrl}discover/movie${jacobKey}`)
    .then(response => response.json())
    .then(result => console.log(result));
}

