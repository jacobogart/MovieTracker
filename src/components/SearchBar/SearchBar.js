import React, { Component } from 'react';

class SearchBar extends Component {
  render() {
    return (
      <div className="SearchBar">
        <i class="fas fa-search"></i>
        <input type="text" className="search-input" placeholder="Search for a movie..."/>
        <i class="fas fa-times"></i>
      </div>
    );
  }
}

export default SearchBar;