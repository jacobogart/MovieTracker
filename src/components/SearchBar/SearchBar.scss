.SearchBar {

  padding: 0 125px;
  background: rgba(white, .1);
  border: rgba(white, .1) solid 3px;

  .search-input {
    background: transparent;
    border: none;
    color: white;
    font-size: 1.1em;
    padding: 15px 0;
    width: 100%;

    &::placeholder {
      color: rgba(white, .3);
    }
  }
}