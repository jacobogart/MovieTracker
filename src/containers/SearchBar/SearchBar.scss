.SearchBar {
  background: rgba(white, .1);
  border: rgba(white, .1) solid 3px;
  direction: rtl;
  display: flex;
  padding: 0 125px;

  .fa-times,
  .fa-search {
    color: rgba(white, .2);
    font-size: 1.2em;
    position: relative;
    top: 12px;
  }

  .search-input {
    background: transparent;
    border: none;
    color: white;
    direction: ltr;
    font-size: 1.1em;
    font-weight: 600;
    margin: 0 0 0 10px;
    padding: 12px 0;
    width: 100%;

    &::placeholder {
      color: rgba(white, .3);
      font-style: italic;
      font-weight: 400;
    }

    &:focus {
      outline: none;
    }

    &:hover + .search-icon > *,
    &:focus + .search-icon > * {
      color: white;
    }
  }

  .clear-search > * {
    width: 20px;

    &:hover {
      cursor: pointer;
      color: white;
    }
  }
}
