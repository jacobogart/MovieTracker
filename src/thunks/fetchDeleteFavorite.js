import { serverUrl } from "../api/pathNames";
import { deleteUserFavorite } from "../actions";

export const fetchDeleteFavorite = (user_id, movie_id) => {
  const url = `${serverUrl}/${user_id}/favorites/${movie_id}`;
  const body = { user_id, movie_id };
  const init = {
    method: 'DELETE',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  };

  return (dispatch) => {
    fetch(url, init)
      .then(response => {
        if (!response.ok) {
          throw Error("Failed to delete favorite");
        } else {
          return response.json();
        }
      })
      .then(result => {
        console.log(result);
        dispatch(deleteUserFavorite(movie_id));
      })
      .catch(error => console.log(error));
    ;
  };
};