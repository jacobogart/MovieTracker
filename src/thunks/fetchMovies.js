import { jacobKey } from '../utils/API-logins';
import { baseUrl } from '../api/pathNames.js';
import { addDiscoverMovies } from '../actions';
import { cleanAllMovies } from '../api/cleaners';

export const fetchMovies = (fetchCase, query) => {
  let url;
  
  switch(fetchCase) {
    case 'search':
      url = `${baseUrl}search/movie${jacobKey}&language=en-US&query=${query}&page=1&include_adult=false`
      break;
    case 'discover':
      url = `${baseUrl}discover/movie${jacobKey}`
      break;
    default:
      console.log('Need fetchCase');
  } 

  return async (dispatch) => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error("Search failed");
      }
      const data = await response.json();
      const movies = cleanAllMovies(data.results);

      dispatch(addDiscoverMovies(movies));
    } catch(error) {
      console.log(error);
    }
  }
}