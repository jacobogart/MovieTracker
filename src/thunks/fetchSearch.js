import { jacobKey } from '../utils/API-logins';
import { baseUrl } from '../api/pathNames.js';
import { addDiscoverMovies } from '../actions';
import { cleanAllMovies } from '../api/cleaners';

export const fetchSearch =  (query) => {
  const url = `${baseUrl}search/movie${jacobKey}&language=en-US&query=${query}&page=1&include_adult=false`
  return async (dispatch) => {
    try {
      const response = await fetch(url);
      if (!response.ok) {
        throw new Error("Search failed");
      }
      const data = await response.json();
      const movies = cleanAllMovies(data.results)
      dispatch(addDiscoverMovies(movies));
    } catch(error) {
      console.log(error);
    }
  }
}