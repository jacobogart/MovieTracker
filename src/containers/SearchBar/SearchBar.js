import React, { Component } from 'react';
import { connect } from 'react-redux';
import { fetchMovies } from '../../thunks/fetchMovies';

class SearchBar extends Component {
  constructor() {
    super();
    this.state = {
      query: ''
    }
  }

  handleSubmit = (e) => {
    e.preventDefault();
    this.props.fetchSearch('search', this.state.query);
  }

  handleChange = (e) => {
    this.setState({ query: e.target.value})
  }

  clearInput = () => {
    this.setState({ query: '' });
  }


  render() {
    return (
      <form 
        className="SearchBar"
        onSubmit={this.handleSubmit}
      >
        <i className="fas fa-search"></i>
        <input 
          type="text" 
          className="search-input" 
          placeholder="Search for a movie..."
          value={this.state.query}
          onChange={this.handleChange}
        />
        <i 
          className="fas fa-times"
          role="button"
          onClick={this.clearInput}
        ></i>
      </form>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  fetchSearch: (fetchCase, query) => dispatch(fetchMovies(fetchCase, query))
})

export default connect(null, mapDispatchToProps)(SearchBar);