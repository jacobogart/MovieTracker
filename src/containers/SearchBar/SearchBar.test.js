import React from 'react';
import { shallow } from 'enzyme';
import { SearchBar, mapDispatchToProps } from "./SearchBar";
import { fetchMovies } from '../../thunks/fetchMovies';

jest.mock('../../thunks/fetchMovies');
const mockFetchSearch = jest.fn();
const mockSubmitEvent = {
  preventDefault: () => {}
};
const mockChangeEvent = {
  target: { value: "The Parent Trap" }
};
const mockHistory = { push: jest.fn() }



describe('SearchBar', () => {
  let wrapper, instance;

  beforeEach(()=> {
    wrapper = shallow(<SearchBar fetchSearch={mockFetchSearch} history={mockHistory}/>);
    instance = wrapper.instance();
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a default state', () => {
    expect(wrapper.state()).toEqual({ query: '' });
  });

  describe('handleChange', () => {
    it('should set state to the input value', () => {
      expect(wrapper.state('query')).toEqual('');
      instance.handleChange(mockChangeEvent);
      expect(wrapper.state('query')).toEqual('The Parent Trap');
    });
  });

  describe('handleSubmit', () => {
    it('should invoke the fetchSearch method', () => {
      instance.handleChange(mockChangeEvent);
      instance.handleSubmit(mockSubmitEvent);
      expect(instance.props.fetchSearch).toHaveBeenCalledWith('search', 'The Parent Trap');
    });
    
    it('should invoke the push method of router history', () => {
      instance.handleSubmit(mockSubmitEvent);
      expect(instance.props.history.push).toHaveBeenCalledWith('/search');
    });
  });

  describe('clearInput', () => {
    it('should reset the query property of state', () => {
      instance.handleChange(mockChangeEvent);
      expect(wrapper.state('query')).toEqual('The Parent Trap');
      instance.clearInput();
      expect(wrapper.state('query')).toEqual('');
    });
  });

  describe('mapDispatchToProps', () => {
    it('should call dispatch with using the fetchSearch method', () => {
      const mockDispatch = jest.fn();
      const thunkToDispatch = fetchMovies('search', 'The Parent Trap');
      const mappedProps = mapDispatchToProps(mockDispatch);
      mappedProps.fetchSearch('search', 'The Parent Trap');
      expect(mockDispatch).toHaveBeenCalledWith(thunkToDispatch);
    });
  });
});