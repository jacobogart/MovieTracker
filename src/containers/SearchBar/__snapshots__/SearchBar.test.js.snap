// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar should match snapshot 1`] = `
<form
  className="SearchBar"
  onSubmit={[Function]}
>
  <div
    className="clear-search"
    role="button"
  >
    <i
      className="fas fa-times"
      onClick={[Function]}
    />
  </div>
  <input
    className="search-input"
    onChange={[Function]}
    placeholder="Search for a movie..."
    type="text"
    value=""
  />
  <div
    className="search-icon"
  >
    <i
      className="fas fa-search"
    />
  </div>
</form>
`;
